base/linalg/dense.jl

```diff
@@ -29,6 +33,41 @@ function scale!(X::Array{T}, s::Real) where T<:BlasComplex
     X
 end
 
+
+function isone(x::StridedMatrix)
+    m, n = size(x)
+    m != n && return false # only square matrices can satisfy x == one(x)
+    if m < div(ISONE_CUTOFF * m * m, sizeof(x))
+        _isone_triacheck(x, m)
+    else
+        _isone_fastcache(x, m)
+    end
+end
+
+@inline function _isone_triacheck(x::StridedMatrix, m::Int)
+    @inbounds for i in 1:m, j in i:m
+        if i == j
+            isone(x[i,i]) || return false
+        else
+            iszero(x[i,j]) && iszero(x[j,i]) || return false
+        end
+    end
+    return true
+end
+
+
+@inline function _isone_fastcache(x::StridedMatrix, m::Int)
+    @inbounds for i in 1:m, j in 1:m
+        if i == j
+            isone(x[i,i]) || return false
+        else
+            iszero(x[j,i]) || return false
```

Comment:
It's fine with me. In my head, `i` is always the variable in the outer scope, so it signals that rows are in the inner loop, but maybe that's just me?